### ISSUE TITLE
mono_domain_set_config

### ISSUE THREAD

**Issue by Numpsy** [ThesaurusSoftware/Dubrovnik#1]
Hi,

While trying to use Dubrovnik with some C# libs that are developed on Windows, I ran into a variant of the issue attached to https://github.com/mono/mono/commit/57f5187ad29a7913f083a659ea77d90eb8bad4d4 (i'm not trying to load any config myself, but i'm seeing the mentioned exception being thrown from the .net serialization lib when it tries to load its own config).

The issue is that the appdomain created by mono_jit_init doesn't have its config bits set up, and the fix is to call mono_domain_set_config() on it before using it. I was wondering if it would be useful to have a Dubrovnik wrapper for the function, but also if it would be worthwhile for the default environment to set it to a useful default given the possible issues with system libs if it isn't done?

Thanks.

**Comment by Thesaurus** [comment]
Hi

That sounds reasonable. At present I am utilising against an older version of Mono that doesn't define mono_domain_set_config(). Once I upgrade (which I need to do) this should be straight forward.

**Comment by Numpsy** [comment]
We started off using an old version, and then had to upgrade for various reasons (including getting the fix for this). I'm actually just getting it set up to use a build of 4.2.1 (seems like I need to pick up the El Capitan fixes).

**Comment by Thesaurus** [comment]
Yes. I am thinking the same thing. Are you running a 32 or 64 bit build of Mono. The current OS X MDK download is still 32 bit I think.

**Comment by Numpsy** [comment]
I'm using a 64bit build that I built myself.

**Comment by Thesaurus** [comment]
Hmm.
Just wondering if you have an insight into getting the current Git mono 4.0.0 branch to build on OS X 10.10 Xcode 6.
I am using the autogen 64 bit build sequence described here http://www.mono-project.com/docs/compiling-mono/mac/

I had success building Mono 3.2.8 some time ago.

make fails like so:
config.status: creating config.h
config.status: config.h is unchanged
Git submodules updated successfully
/Applications/Xcode.app/Contents/Developer/usr/bin/make  all-recursive
Making all in po
Making all in mcs
make[3]: Nothing to be done for `all-am'.
Making all in libgc
Making all in m4
make[3]: *** No rule to make target `source/mono/libgc/missing makeinfo '.  Stop.
make[2]: *** [all-recursive] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2

**Comment by Thesaurus** [comment]
Sorry for the noise.
There was a space in one of the folder names leading to the source files.
I wonder what idiot put that there...

**Comment by Thesaurus** [comment]
I have added a convenience method for mono_domain_set_config(). We can set this to a suitable path by default. I don't require this call myself so I am a bit fuzzy as to what to set the daft to. I presume that the managed layer needs read write access to the location and that it must be created if it doesn't exist.

**Comment by Numpsy** [comment]
I don't currently need to read/write any explicit settings myself, for now I just need the config machinery set up to avoid some problems when the mono libs themselves try to use it (and I think it may be ok if the file doesn;t exist when they only want to read it). If i do need to write things at some point, i'll probably set it to some app specific location.

When i was looking for info on this issue I saw some other callers point it at the machine config file at /Library/Frameworks/Mono64.framework/Versions/version/etc/mono/4.5/machine.config by default, which may be the sensible default? If not, maybe the directory the current app is running from?

**Comment by Numpsy** [comment]
on a related fyi, mono 4.0 will crash if you call domain_set_config with a config file name shorter than ".config" (fixed in 4.2 @ https://github.com/mono/mono/commit/ac4cf68d878925a635c93c3c30da7f02c04a23a3)

**Comment by Thesaurus** [comment]
My last commit to implement mono_domain_set_config was utter crap. Current commit should improve things.

By default the config file is loaded from /Library/Frameworks/Mono64.framework/Versions/Current/etc. If you are running Mono from outside  /Library/Frameworks call DBManagedEnvironment -+ (void)configureAssemblyRootPath:(NSString *)monoAssemblyRootFolder configRootFolder:(NSString *)monoConfigFolder.

A normal user process certainly won't be able to write to /Library/Frameworks.
I am not quite sure under what circumstances you need to call mono_domain_set_config(). Perhaps ask on the Mono list for further clarification.

**Comment by Numpsy** [comment]
I'll update my build and give it a go.

I ran into problems in this area initialliy when I started getting exceptions thrown from the static constructor for XmlSerializer (https://github.com/mono/mono/blob/master/mcs/class/System.XML/System.Xml.Serialization/XmlSerializer.cs#L155), and this seems to be the 'fix' for that.
I just need to keep an eye on it I guess.

**Comment by Thesaurus** [comment]
If you look at the project README you can see that I have added a build script to ease building the Mono framework on OS X. I am currently targeting Mono 4.0.4.4.

**Comment by Numpsy** [comment]
> > I am currently targeting Mono 4.0.4.4.

Lucky me has just started submitting stubs for a few unimplemented functions in the Mono class libs. No idea what version that will end up in, but it's a learing experience at least ;-)

**Comment by Thesaurus** [comment]
I saw your exchange on the Mono list.
4.2 is in Alpha - http://www.mono-project.com/docs/about-mono/releases/4.2.0/.
Looks like this will pull in more MS reference source.
Might be worth checking if the functions you are targeting are now available in the reference sources.

Good luck!
